[RFC] Travis: Install Python 3 Neovim module.

fwalch opened the pull request:
Supersedes #3152.

dnmario commented:
Dear Florian Walch,

The email sent to danielm@tinyhub.tk with subject
"[neovim] [RFC] Travis: Install Python 3 Neovim module. (#3316)"
will be delivered as soon as possible.

Your email address is automatically added to the
recipients 'whitelist'. This means for further
emails to danielm@tinyhub.tk, you will not receive a delivery
notification again.

TK MAILIAS
*************************************************
Looking for a way to Protect Your Identity and 
for a great no-more-spam solution for your email
address reply@reply.github.com? Sign up for TK Mailias.

It's free! Please visit http://www.mail.tk/

fwalch commented:
@dnmario You might want to check the configuration of your mail account.

dnmario commented:
Dear Florian Walch,

The email sent to danielm@tinyhub.tk with subject
"Re: [neovim] [RFC] Travis: Install Python 3 Neovim module. (#3316)"
will be delivered as soon as possible.

Your email address is automatically added to the
recipients 'whitelist'. This means for further
emails to danielm@tinyhub.tk, you will not receive a delivery
notification again.

TK MAILIAS
*************************************************
Looking for a way to Protect Your Identity and 
for a great no-more-spam solution for your email
address reply@reply.github.com? Sign up for TK Mailias.

It's free! Please visit http://www.mail.tk/

justinmk commented:
LGTM!

dnmario commented:
Dear Justin M. Keyes,

The email sent to danielm@tinyhub.tk with subject
"Re: [neovim] [RFC] Travis: Install Python 3 Neovim module. (#3316)"
will be delivered as soon as possible.

Your email address is automatically added to the
recipients 'whitelist'. This means for further
emails to danielm@tinyhub.tk, you will not receive a delivery
notification again.

TK MAILIAS
*************************************************
Looking for a way to Protect Your Identity and 
for a great no-more-spam solution for your email
address reply@reply.github.com? Sign up for TK Mailias.

It's free! Please visit http://www.mail.tk/

justinmk commented:
The good news is that the python3 tests are now running :) Bad news is there is something wrong with them :P

fwalch commented:
Looks like the Python 3 module is installed somewhere, as `python3 -c 'import neovim'` worked and caused the test suite to run python3 spec, but it still can't be detected by Neovim's provider, which is why the spec failed:

```
Python 3 (or the Python 3 neovim module) is broken or missing:
provider/pythonx: Could not load Python 3:
/usr/bin/python3 does have not have the neovim module installed. See ":help nvim-python".
```

fwalch commented:
With #3351:

```
provider/pythonx: Could not load Python 3:
/usr/bin/python3 is Python 3.2Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'module' object has no attribute 'find_loader'
 and cannot provide Python >= 3.3.
```

fwalch commented:
I played around with setting `language: python`, but I think it's not feasible to use that here.. I think we have to wait until Travis gets a Python 3 >= 3.3 or somehow set up an own virtualenv.

justinmk commented:
Foiled at every turn :/

fwalch commented:
Okay... I found a `python3.5` in one of the unlocked PPAs (I tried using the Debian Sid repo, but that messed up everything). There's no accompanying pip or easy_install package, so I'm now trying to install pip manually.

ZyX-I commented:
@fwalch Powerline compiles its own Python in its bot-ci: https://github.com/powerline/bot-ci/blob/842ff93f9f5d84f6cc2921567146604608a9d82b/scripts/build-ucs2-cpython.sh, https://github.com/ZyX-I/powerline/blob/5ebf0875b85f85e7cd4c1e07aa8cdebbcc3bd17e/tests/install.sh#L32-L48. And this works in containers, despite not being a package (and thus whitelisted).

justinmk commented:
Will compiling python add much time to the builds? Perhaps we should wait for travis to whitelist python 3.3.

ZyX-I commented:
@justinmk This is rather lengthy (especially given that after compiling Python my bot also compiles Vim and zsh+zpython). But check e.g. https://travis-ci.org/powerline/bot-ci/jobs/69198544: it says it took only 13 minutes (side note: I completely do not trust travis times). Cannot say how long it will take for Python-3\* because Python-3\* cannot be compiled with UCS2 unicode strings (and thus needs not be tested), but do not forget that 13 minutes is for cpython+zsh+zpython+vim (two times: v7.0.112 and newest version)+creating wheel packages.

ZyX-I commented:
Also maybe Python is still available in `/opt`, but is not added to `$PATH`. Better to check this first, setting up own virtualenv is not much of a problem.

fwalch commented:
@ZyX-I Thanks! I'll wait and see if Travis whitelists the python3.5 package soon, otherwise I'll try the other options.

fwalch committed:
test: Check for installed Python module with Vimscript function.

Use the existing Vimscript function provider#pythonx#Detect()
to determine whether the Neovim Python module is installed and
Python 2/3 tests can be run.

fwalch committed:
Travis: Install Python 3 to run Python 3 tests.

Install Python 3.3 from the Deadsnakes PPA. As this doesn't have pip,
install it manually into ~/.local.

~/.local/bin is apparently in Travis's default PATH, meaning "pip"
doesn't refer to Python 2's pip anymore, but to the manually
installed Python 3 version. Updated the scripts to use version-
suffixed executable names (e.g. pip2.7).

Set CC=cc to use system's default compiler when installing Python
modules, as gcc on OS X had a problem with compiling one of the
dependencies of the Neovim Python module.

fwalch commented:
Okay, I believe this should work now :-)

```
# Linux and OSX
598 successes / 0 failures / 0 errors / 3 pending
```

With the first commit, a missing Python module will cause output like:

```
Pending → ...rojects/neovim/test/functional/provider/python3_spec.lua @ 10
Python 3 (or the Python 3 neovim module) is broken or missing:
provider/pythonx: Could not load Python 3:
/usr/bin/python3 does have not have the neovim module installed. See ":help nvim-python".
python3.5 not found in search path or not executable.
/usr/bin/python3.4 does have not have the neovim module installed. See ":help nvim-python".
python3.3 not found in search path or not executable.
/usr/bin/python does have not have the neovim module installed. See ":help nvim-python".
```

fwalch commented:
Something is wrong with coveralls.. output is only:

```
Usage:
    coveralls [options]
    coveralls debug  [options]
```

Trying to debug this.

fwalch commented:
Errors on OSX, but only for GCC build:

```
Error -> ...d/neovim/neovim/test/functional/provider/python_spec.lua @ 18
python commands and functions before_each
./test/functional/helpers.lua:67: Vim(finish):E168: :finish used outside of a sourced file
stack traceback:
	./test/functional/helpers.lua:67: in function 'request'
	./test/functional/helpers.lua:125: in function 'command'
	...d/neovim/neovim/test/functional/provider/python_spec.lua:20: in function <...d/neovim/neovim/test/functional/provider/python_spec.lua:18>
```

Restarting the job doesn't help.

fwalch commented:
This should be it, finally. gcc on OSX could not compile one of the dependency Python modules (I think msgpack), but the build nevertheless proceeded. Working around that by setting `CC=cc` when doing `pip install` to use the system's default compiler (added a note to the code and to the commit message).

fwalch commented:
Rebased.

@justinmk Any more comments? Otherwise I think this can be merged.

justinmk commented on test/functional/provider/python_spec.lua:
```diff
@@ -1,19 +1,19 @@
-  local proc = io.popen(
-    [[python -c 'import neovim, sys; sys.stdout.write("ok")' 2> /dev/null]])
-  if proc:read() ~= 'ok' then
+  command('let [interp, errors] = provider#pythonx#Detect(2)')
```
This is setting globals `g:interp`, `g:errors`. Would be better to be explicit about that. The names also seem a bit risky for globals, something like `g:pydetect_errors` would be less risky.

justinmk replied:
Although it shouldn't matter after `clear()`, I guess.

justinmk commented:
How much time does this add to the build? Can you mark the places which are to be removed after travis gets python 3.x?

fwalch commented:
Regarding the build times: That's really hard to say. Looking at two recent builds on master, one `clint` build took 1min28s, the other 2min43s. One GCC build took 4min6s, the other 5min26s. This one took 4min20s. So.. I guess it's negligible.

I already had a TODO in the .travis.yml to remove the Python PPA, added another one in before_install.sh to remove pip installation if we get a python3-pip APT package.

justinmk commented:
Ok. I'd feel better if the viml globals were at least `g:` prefixed. Otherwise LGTM

fwalch commented:
Okay, I'll change that and then merge.

justinmk commented:
@fwalch There is a problem with [master build](https://travis-ci.org/neovim/neovim/jobs/82420476):

```
Setting up build cache
$ export CASHER_DIR=$HOME/.casher
0.03s$ Installing caching utilities
1.02sattempting to download cache archive
fetching master/cache--compiler-clang-3.6.tgz
found cache
1.76sadding /home/travis/.cache/pip to cache
adding /home/travis/.cache/nvim-deps to cache
$ clang-3.6 --version
/home/travis/build.sh: line 41: clang-3.6: command not found
1.98s$ .ci/before_install.sh
Upgrade Python 2's pip.
You are using pip version 6.0.8, however version 7.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Collecting pip from https://pypi.python.org/packages/py2.py3/p/pip/pip-7.1.2-py2.py3-none-any.whl#md5=5ff9fec0be479e4e36df467556deed4d
  Using cached pip-7.1.2-py2.py3-none-any.whl
Installing collected packages: pip
Successfully installed pip-6.0.8
Install pip for Python 3.
.ci/before_install.sh: line 27: python3.3: command not found
```

fwalch commented:
@justinmk looks like https://github.com/neovim/neovim/pull/2506#issuecomment-143070433

justinmk commented:
@fwalch Thanks, restart fixed it.
